### ISSUE TITLE
Heterogeneity in terms of network and device performance

### ISSUE THREAD

**Issue by ekr** [IRTF-HRPC/draft-guidelines#2]
S 4.1 says:
>   Also considering the fact that network quality and conditions vary
   across geography and time, it is also important to design protocols
   such that they are reliable even on low bandwidth and high latency
   connections.

And then:
>    Example: Heterogeneity is inevitable and needs be supported by
   design.  As far as possible, multiple types of hardware must be
   allowed for (e.g., transmission speeds differing by at least 7 orders
   of magnitude, various computer word lengths, and hosts ranging from
   memory-starved microprocessors up to massively parallel
   supercomputers).  Multiple types of application protocols must be
   allowed for, ranging from the simplest such as remote login up to the
   most complex such as commit protocols for distributed databases.
   [RFC1958].

As noted, we routinely design systems that in fact only work well on fairly capable machines and networks and importantly *make decisions* about protocol design that trade off working on some classes of devices in favor o other values, for instance by making routine use of encryption, even if that's too expensive for some systems. So, I think "As far as possible" is way too strong. This is one of many factors to weigh.

**Comment by nllz** [comment]
Sure, it's one of the considerations, and it is listed as such. For context, we added "as far as possible" in response to your initial review.

There are many other guidelines in the document, and all of them are voluntary (such is the nature of guidelines), and simply ways to think through design decisions and their impact. If it makes sense to design protocols to only work on particular kinds of machines, this choice can (and will) obivously be made (we are not the protocol police, this document is informational, and we are an RG), but it would be very good if people make this particular choice, that they document their reasons for doing so. 

We could add:
    If you choose to not support general purpose computing hardware, it would be good to document the choices that led to this decision.

**Comment by ekr** [comment]
> Sure, it's one of the considerations, and it is listed as such. For context, we added "as far as possible" in response to your initial review.

Yes, and as I noted just above, I think it's way too strong

> There are many other guidelines in the document, and all of them are voluntary (such is the nature of guidelines),

"must" is not a word I typically associate with "guidelines"

> If you choose to not support general purpose computing hardware, it would be good to document the choices that led to this decision.

The issue is not *general purpose computing hardware* but rather this specific text:

"e.g., transmission speeds differing by at least 7 orders
of magnitude, various computer word lengths, and hosts ranging from
memory-starved microprocessors up to massively parallel
supercomputers"

I don't think it's incumbent upon protocol designers to explain why,
for instance, they decided to design a protocol that won't work on
an Intel 4004.

**Comment by nllz** [comment]
s/must/should then?

**Comment by ekr** [comment]
I don't think this document should take a normative position for the reasons I indicated above.

**Comment by nllz** [comment]
OK, so I removed the normative language, does this work:

Example: Heterogeneity is inevitable and needs be supported by design. As far as possible, multiple types of hardware must be allowed for (e.g., transmission speeds differing by at least 7 orders of magnitude, various computer word lengths, and hosts ranging from memory-starved microprocessors up to massively parallel supercomputers). What made the internet successful is its ability to accommodate for many different kinds of protocols and devices, ranging from the simplest such as remote login up to the most complex such as commit protocols for distributed databases. [RFC1958].

**Comment by ekr** [comment]
Heterogeneity is inevitable and needs be supported by design. As
   far as possible, multiple types of hardware must be allowed for
   (e.g., transmission speeds differing by at least 7 orders of
   magnitude, various computer word lengths, and hosts ranging from
   memory-starved microprocessors up to massively parallel
   supercomputers). What made the internet successful is its ability
   to accommodate for many different kinds of protocols and devices,
   ranging from the simplest such as remote login up to the most
   complex such as commit protocols for distributed
   databases. [RFC1958].


I think this needs to be unpacked some. Specifically, I think
we need to distinguish between the network substrate and endpoints.

I somewhat agree with the claim that the basic design of IP (and TCP)
and its ability to run on any substrate is a big contributor to the
Internet's success, for the specific reason that you can run one
protocol and not two or more. I do think that saying that this is
"what made the Internet successful" is too strong, as there are a lot
of factors in play, but I don't disagree that this kind of flexibility
is important.

I think this argument is a lot weaker the higher we get in the stack.
For example:

* While it's true that the Web does a good job of covering a somewhat
  broad range of devices (though not really 7 orders of magnitude) in my
  experience supporting older devices is often considered a burden, and
  it's quite common to have Web sites which just don't work well on old
  devices. It's true that HTTP itself actually holds up pretty well with
  less capable devices (though I know this is debated) but the Web
  platform and other protocols such as WebRTC maybe not so much.

* This isn't my main area of expertise, but my sense is that
  constrained devices don't need to really run fancy routing
  protocols like BGP, etc. Is it really important that I be
  able to run BGP on my watch?

* There has of course been a lot of concern about the impact of
  cryptographic protocols on low-end devices, hence LAKE, cTLS,
  etc.

Taking the example of the Web again, ISTM that the value here is that
people who have limited resources to be able to use the Web, but
that's not the same as saying that it needs to run on a PIC
microcontroller. Rather, at most it should work on the kind of limited
devices that people have in practice, which are actually quite
capable. For instance KaiOS devices seem to have around 4GB of memory
[0]. And of course, if we were to say that the Web had to work on very
limited devices, this would have a fairly serious impact on the
experience of the Web for everyone else.

TBH, I'm having a bit of trouble thinking of a case where this
text or something like it would have been a useful guide to protocol
design. Do you have one?


[0] https://www.kaiostech.com/explore/devices/

**Comment by nllz** [comment]
The text is a quote from RFC1958 - which should probably have been better indicated with double quotes (!)

So I made some changes, that probably make the whole thing a bit more palatable:

Example: Heterogeneity significantly contributed to the success of the internet architecture (https://dash.harvard.edu/handle/1/4455262). Niels Bohr famously said: "Prediction is very difficult, especially if it’s about the future", this also holds true for future uses of the internet architecture and infrastructure. Therefore, as a rule of thumb it is important to  - as far as possible - design your protocol for different devices and uses. RFC1958 suggests: transmission speeds differing by at least 7 orders of magnitude, various computer word lengths, and hosts ranging from memory-starved microprocessors up to massively parallel supercomputers). If you choose not to do this, it could be relevant to document the reasoning for that.

**Comment by ekr** [comment]
I'm sorry, but this doesn't really make it more palatable. 

To try to summarize my point, my basic objection is that I don't think that trying to get people to design for this wide range of capabilities is actually very good advice under most circumstances, and so I don't think this document should give it.  I recognize that you've tried to weaken the normative implications of the text, but my objection is primarily to the 
advice itself, at least the way it's framed here.

At this point, I think rather than trying to wordsmith this text into some kind of compromise, it would be more useful to try to actually reach a common understanding. My comment above makes a fairly extensive argument for the position I'm espousing here. What part of this do you think I have wrong?

**Comment by nllz** [comment]
I'd argue that the value of the internet and all connected devices increases as it can be used as a general-purpose network. This facilitated innovation and seeks to limit the impact of path-dependency. Furthermore, it serves people who use devices with different capacities and capabilities. It also ensures that devices can be re-used for other uses, and thus limiting e-waste. There might indeed be reasons to deviate from this, but if people do that, it would be good if they document as to why they do that. This relates to outcome transparency.

In your first post you write: "So, I think "As far as possible" is way too strong. This is one of many factors to weigh.". I don't disagree, and if people already do this balancing that is great, if they don't we could help them think about this through this text. IMHO it would be great if this would be documented. 

I can also write an extensive part about ossification vs heterogeneity if that would convince you.

**Comment by nllz** [comment]
In your comment from May 14 you write that the text  "e.g., transmission speeds differing by at least 7 orders
of magnitude, various computer word lengths, and hosts ranging from memory-starved microprocessors up to massively parallel supercomputers" is your actual problem with this part, but this is simply straight from RFC1958, so am not sure we can do a lot about that. Perhaps better to take that up with the author of that RFC ?

**Comment by ekr** [comment]
A few points here:

RFC 1958 is not an IETF consensus document, so I think this document is free to endorse it or not. The text you propose clearly takes a position endorsing it. So while I might or might not disagree with 1958, the relevant question is what this document  says.

Re: the Internet as a general purpose network, as I said in my comment above, I think that applies much more strongly to lower-level protocols than to application protocols, and that trying to have application protocols operate across a very broad range of computing devices is actually not really that great an idea, because it forces those protocols into what fits in the least capable devices. I do think that protocols should not unnecessarily consume resources, but that's an engineering judgement call of the kind we routinely make.

IETF protocols do not generally come with rationale sections for each design decision, and I don't think that this document or your comment above make a persuausive case that this specific decision should be different.

**Comment by nllz** [comment]
I did not write that "application protocols [have to] operate across a very broad range of computing devices". The only thing I am saying is that the general purpose (or in the words of Zittrain 'generative') nature of the internet has been one of its main success factors, and that if we deviate from that, it would be good to document it. 

If you disagree I am afraid this is an 'agree to disagree'.

**Comment by ekr** [comment]
Well, I don't disagree with the statement you are making here, but (1) that's not what the text here or that you proposed says and (2) it's something that really only applies to certain classes of protocol and (3) there already is text elsewhere in S 4.8 that talks about general purpose.

**Comment by nllz** [comment]
OK, let me try again then. Would this work (removed text from RFC1958 and added higher relevance for lower layers)

Example: Heterogeneity significantly contributed to the success of the internet architecture https://dash.harvard.edu/handle/1/4455262). Niels Bohr famously said: "Prediction is very difficult, especially if it’s about the future", this also holds true for future uses of the internet architecture and infrastructure. Therefore, as a rule of thumb it is important to - as far as possible - design your protocol for different devices and uses, especially at lower layers of the stack. However, if you choose not to do this, it could be relevant to document the reasoning for that.

**Comment by nllz** [comment]
https://github.com/IRTF-HRPC/draft-guidelines/commit/0aa8898b8a1097a5257b11c660285e8d8c9f9f62

**Comment by nllz** [comment]
I propose we move the discussion back to the hrpc list based on: https://datatracker.ietf.org/doc/html/draft-irtf-hrpc-guidelines-19

**Comment by ekr** [comment]
I still think this text is pretty muddled, but in the interest of moving forward,I can live with it. Authors, feel free to close.
